R/PipeOpEncodePL.R

```diff
@@ -0,0 +1,409 @@
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpEncodePL`]/[`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @references
+#' `r format_bib("gorishniy_2022")`
+#'
+#' @family PipeOps
+#' @family Piecewise Linear Encoding PipeOps
+#' @template seealso_pipeopslist
+#' @include PipeOpTaskPreproc.R
+#' @export
+#' @examples
+#' library(mlr3)
+#'
+#' # For classification task
+#' task = tsk("iris")$select(c("Petal.Width", "Petal.Length"))
+#' pop = po("encodepltree", task_type = "TaskClassif")
+#' train_out = pop$train(list(task))[[1L]]
+#'
+#' # Calculated bin boundaries per feature
+#' pop$state$bins
+#' # Each feature was split into three encoded features using piecewise linear encoding
+#' train_out$head()
+#'
+#' # Prediction works the same as training, using the bins learned during training
+#' predict_out = pop$predict(list(task))[[1L]]
+#' predict_out$head()
+#'
+#' # Controlling behavior of the tree learner, here: setting minimum number of
+#' # observations per node for a split to be attempted
+#' pop$param_set$set_values(minsplit = 5)
+#'
+#' train_out = pop$train(list(task))[[1L]]
+#' # feature "hp" now gets split into five encoded features instead of three
+#' pop$state$bins
+#' train_out$head()
+#'
+#' # For regression task
+#' task = tsk("mtcars")$select(c("cyl", "hp"))
+#' pop = po("encodepltree", task_type = "TaskRegr")
+#' train_out = pop$train(list(task))[[1L]]
+#'
+#' # Calculated bin boundaries per feature
+#' pop$state$bins
+#' # First feature was split into three encoded features, second into two, using piecewise linear encoding
+#' train_out$head()
+PipeOpEncodePLTree = R6Class("PipeOpEncodePLTree",
+  inherit = PipeOpEncodePL,
+  public = list(
+    initialize = function(task_type, id = "encodepltree", param_vals = list()) {
+      assert_choice(task_type, mlr_reflections$task_types$task)
+      if (task_type == "TaskRegr") {
+        private$.tree_learner = LearnerRegrRpart$new()
+      } else if (task_type == "TaskClassif") {
+        private$.tree_learner = LearnerClassifRpart$new()
+      } else {
+        stopf("Task type %s not supported.", task_type)
+      }
+
+      super$initialize(id, param_set = alist(private$.tree_learner$param_set), param_vals = param_vals,
+        packages = private$.tree_learner$packages, task_type = task_type)
+    }
+  ),
+  private = list(
+
+    .tree_learner = NULL,
+
+    .get_bins = function(task, cols) {
+      learner = private$.tree_learner
+      bins = list()
+      for (col in cols) {
+        t = task$clone(deep = TRUE)$select(col)
+        model = learner$train(t)$model
+        frame = model$frame
+
+        # Extracting splits according to https://github.com/cran/rpart/blob/983544575b80df286bf8ce238faf4afa145872cd/R/labels.rpart.R#L26
+        # NOTE: Maybe this isn't necessary, and we can just take index of splits (should splits exist), since compete and
+        # surrogate *might* only be larger than 0 than trained with more than one feature?
```

Comment:
```suggestion
        # NOTE: `frame$nsurrogate` and `frame$ncompete` seem to be always 0 in our case, since competing and
        # surrogate splits only exist when there are more than one freature. Leaving it in for completeness.
```